// chat-backend/dist/modules/messages/messages.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMessages = getMessages;
exports.sendMessage = sendMessage;
exports.editMessage = editMessage;
exports.deleteMessage = deleteMessage;
const prisma_1 = require("../../lib/prisma");
const error_middleware_1 = require("../../common/middleware/error.middleware");
const client_1 = require("@prisma/client");
const MESSAGE_PAGE_SIZE = 50;
const msgSelect = {
    id: true, roomId: true, content: true, isEdited: true,
    createdAt: true, updatedAt: true,
    sender: { select: { id: true, username: true } },
    replyTo: {
        select: {
            id: true, content: true,
            sender: { select: { id: true, username: true } },
        },
    },
    attachments: true,
};
async function getMessages(roomId, userId, cursor) {
    const member = await prisma_1.prisma.roomMember.findUnique({
        where: { roomId_userId: { roomId, userId } },
    });
    if (!member)
        throw (0, error_middleware_1.createError)('Not a member', 403);
    const messages = await prisma_1.prisma.message.findMany({
        where: { roomId },
        select: msgSelect,
        orderBy: { createdAt: 'desc' },
        take: MESSAGE_PAGE_SIZE,
        ...(cursor ? { cursor: { id: cursor }, skip: 1 } : {}),
    });
    return messages.reverse();
}
async function sendMessage(roomId, senderId, content, replyToId) {
    const member = await prisma_1.prisma.roomMember.findUnique({
        where: { roomId_userId: { roomId, userId: senderId } },
    });
    if (!member)
        throw (0, error_middleware_1.createError)('Not a member', 403);
    if (content.length > 3072)
        throw (0, error_middleware_1.createError)('Message too long (max 3KB)', 400);
    return prisma_1.prisma.message.create({
        data: { roomId, senderId, content, replyToId },
        select: msgSelect,
    });
}
async function editMessage(messageId, userId, content) {
    const message = await prisma_1.prisma.message.findUnique({ where: { id: messageId } });
    if (!message)
        throw (0, error_middleware_1.createError)('Message not found', 404);
    if (message.senderId !== userId)
        throw (0, error_middleware_1.createError)('Cannot edit others messages', 403);
    if (content.length > 3072)
        throw (0, error_middleware_1.createError)('Message too long', 400);
    return prisma_1.prisma.message.update({
        where: { id: messageId },
        data: { content, isEdited: true },
        select: msgSelect,
    });
}
async function deleteMessage(messageId, userId) {
    const message = await prisma_1.prisma.message.findUnique({ where: { id: messageId } });
    if (!message)
        throw (0, error_middleware_1.createError)('Message not found', 404);
    if (message.senderId !== userId) {
        // Check if user is admin in the room
        const member = await prisma_1.prisma.roomMember.findUnique({
            where: { roomId_userId: { roomId: message.roomId, userId } },
        });
        if (!member || member.role !== client_1.MemberRole.ADMIN) {
            throw (0, error_middleware_1.createError)('Not authorized to delete this message', 403);
        }
    }
    await prisma_1.prisma.message.delete({ where: { id: messageId } });
}
