// chat-backend/dist/common/middleware/upload.middleware.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.uploadFile = exports.uploadImage = void 0;
const multer_1 = __importDefault(require("multer"));
const path_1 = __importDefault(require("path"));
const uuid_1 = require("uuid");
const env_1 = require("../../config/env");
function getStorage(subdir) {
    return multer_1.default.diskStorage({
        destination: (_req, _file, cb) => {
            cb(null, path_1.default.join(env_1.config.uploadsDir, subdir));
        },
        filename: (_req, file, cb) => {
            const ext = path_1.default.extname(file.originalname);
            cb(null, `${(0, uuid_1.v4)()}${ext}`);
        },
    });
}
function imageFilter(_req, file, cb) {
    if (file.mimetype.startsWith('image/')) {
        cb(null, true);
    }
    else {
        cb(new Error('Only image files allowed'));
    }
}
exports.uploadImage = (0, multer_1.default)({
    storage: getStorage('images'),
    limits: { fileSize: env_1.config.maxImageSizeMb * 1024 * 1024 },
    fileFilter: imageFilter,
});
exports.uploadFile = (0, multer_1.default)({
    storage: getStorage('files'),
    limits: { fileSize: env_1.config.maxFileSizeMb * 1024 * 1024 },
});
