// chat-backend/dist/modules/auth/auth.controller.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerHandler = registerHandler;
exports.loginHandler = loginHandler;
exports.logoutHandler = logoutHandler;
exports.getSessionsHandler = getSessionsHandler;
exports.revokeSessionHandler = revokeSessionHandler;
exports.changePasswordHandler = changePasswordHandler;
exports.deleteAccountHandler = deleteAccountHandler;
exports.meHandler = meHandler;
const authService = __importStar(require("./auth.service"));
async function registerHandler(req, res, next) {
    try {
        const { email, username, password } = req.body;
        const user = await authService.register(email, username, password);
        res.status(201).json({ user });
    }
    catch (err) {
        next(err);
    }
}
async function loginHandler(req, res, next) {
    try {
        const { email, password } = req.body;
        const ip = req.ip;
        const browserInfo = req.headers['user-agent'];
        const { token, user } = await authService.login(email, password, ip, browserInfo);
        res.cookie('token', token, {
            httpOnly: true,
            secure: process.env.NODE_ENV === 'production',
            sameSite: 'lax',
            maxAge: 7 * 24 * 60 * 60 * 1000,
        });
        res.json({ user, token });
    }
    catch (err) {
        next(err);
    }
}
async function logoutHandler(req, res, next) {
    try {
        await authService.logout(req.sessionId);
        res.clearCookie('token');
        res.json({ message: 'Logged out' });
    }
    catch (err) {
        next(err);
    }
}
async function getSessionsHandler(req, res, next) {
    try {
        const sessions = await authService.getSessions(req.userId);
        res.json({ sessions });
    }
    catch (err) {
        next(err);
    }
}
async function revokeSessionHandler(req, res, next) {
    try {
        await authService.revokeSession(req.params.sessionId, req.userId);
        res.json({ message: 'Session revoked' });
    }
    catch (err) {
        next(err);
    }
}
async function changePasswordHandler(req, res, next) {
    try {
        const { oldPassword, newPassword } = req.body;
        await authService.changePassword(req.userId, oldPassword, newPassword);
        res.json({ message: 'Password changed' });
    }
    catch (err) {
        next(err);
    }
}
async function deleteAccountHandler(req, res, next) {
    try {
        await authService.deleteAccount(req.userId);
        res.clearCookie('token');
        res.json({ message: 'Account deleted' });
    }
    catch (err) {
        next(err);
    }
}
async function meHandler(req, res, next) {
    try {
        const user = await Promise.resolve().then(() => __importStar(require('../../lib/prisma'))).then(m => m.prisma.user.findUnique({
            where: { id: req.userId },
            select: { id: true, email: true, username: true, createdAt: true },
        }));
        if (!user) {
            res.status(404).json({ error: 'User not found' });
            return;
        }
        res.json({ user });
    }
    catch (err) {
        next(err);
    }
}
