// chat-backend/dist/modules/auth/auth.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = register;
exports.login = login;
exports.logout = logout;
exports.getSessions = getSessions;
exports.revokeSession = revokeSession;
exports.changePassword = changePassword;
exports.deleteAccount = deleteAccount;
const bcrypt_1 = __importDefault(require("bcrypt"));
const prisma_1 = require("../../lib/prisma");
const jwt_1 = require("../../common/utils/jwt");
const error_middleware_1 = require("../../common/middleware/error.middleware");
const SALT_ROUNDS = 12;
async function register(email, username, password) {
    const existingEmail = await prisma_1.prisma.user.findUnique({ where: { email } });
    if (existingEmail)
        throw (0, error_middleware_1.createError)('Email already in use', 409);
    const existingUsername = await prisma_1.prisma.user.findUnique({ where: { username } });
    if (existingUsername)
        throw (0, error_middleware_1.createError)('Username already taken', 409);
    const passwordHash = await bcrypt_1.default.hash(password, SALT_ROUNDS);
    const user = await prisma_1.prisma.user.create({
        data: { email, username, passwordHash },
        select: { id: true, email: true, username: true, createdAt: true },
    });
    return user;
}
async function login(email, password, ip, browserInfo) {
    const user = await prisma_1.prisma.user.findUnique({ where: { email } });
    if (!user)
        throw (0, error_middleware_1.createError)('Invalid credentials', 401);
    const valid = await bcrypt_1.default.compare(password, user.passwordHash);
    if (!valid)
        throw (0, error_middleware_1.createError)('Invalid credentials', 401);
    const expiresAt = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);
    const session = await prisma_1.prisma.session.create({
        data: {
            userId: user.id,
            token: '',
            ip,
            browserInfo,
            expiresAt,
        },
    });
    const token = (0, jwt_1.signToken)({ userId: user.id, sessionId: session.id });
    await prisma_1.prisma.session.update({
        where: { id: session.id },
        data: { token },
    });
    return {
        token,
        user: { id: user.id, email: user.email, username: user.username },
    };
}
async function logout(sessionId) {
    await prisma_1.prisma.session.delete({ where: { id: sessionId } });
}
async function getSessions(userId) {
    return prisma_1.prisma.session.findMany({
        where: { userId, expiresAt: { gt: new Date() } },
        select: { id: true, browserInfo: true, ip: true, createdAt: true },
        orderBy: { createdAt: 'desc' },
    });
}
async function revokeSession(sessionId, userId) {
    const session = await prisma_1.prisma.session.findUnique({ where: { id: sessionId } });
    if (!session || session.userId !== userId)
        throw (0, error_middleware_1.createError)('Session not found', 404);
    await prisma_1.prisma.session.delete({ where: { id: sessionId } });
}
async function changePassword(userId, oldPassword, newPassword) {
    const user = await prisma_1.prisma.user.findUnique({ where: { id: userId } });
    if (!user)
        throw (0, error_middleware_1.createError)('User not found', 404);
    const valid = await bcrypt_1.default.compare(oldPassword, user.passwordHash);
    if (!valid)
        throw (0, error_middleware_1.createError)('Current password is incorrect', 400);
    const passwordHash = await bcrypt_1.default.hash(newPassword, SALT_ROUNDS);
    await prisma_1.prisma.user.update({ where: { id: userId }, data: { passwordHash } });
}
async function deleteAccount(userId) {
    // Cascade deletes owned rooms (and their messages/files via DB cascade)
    await prisma_1.prisma.user.delete({ where: { id: userId } });
}
