// chat-backend/dist/modules/rooms/rooms.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRoom = createRoom;
exports.getPublicRooms = getPublicRooms;
exports.getUserRooms = getUserRooms;
exports.getRoomById = getRoomById;
exports.joinRoom = joinRoom;
exports.leaveRoom = leaveRoom;
exports.deleteRoom = deleteRoom;
exports.updateRoom = updateRoom;
exports.getRoomMembers = getRoomMembers;
exports.banMember = banMember;
exports.unbanMember = unbanMember;
exports.getBannedUsers = getBannedUsers;
exports.makeAdmin = makeAdmin;
exports.removeAdmin = removeAdmin;
exports.removeMember = removeMember;
exports.inviteUser = inviteUser;
exports.acceptInvitation = acceptInvitation;
exports.getPendingInvitations = getPendingInvitations;
exports.getOrCreateDM = getOrCreateDM;
const prisma_1 = require("../../lib/prisma");
const error_middleware_1 = require("../../common/middleware/error.middleware");
const client_1 = require("@prisma/client");
const roomSelect = {
    id: true, name: true, description: true, visibility: true,
    ownerId: true, isDirect: true, createdAt: true,
    _count: { select: { members: true } },
};
async function createRoom(ownerId, name, description, visibility) {
    const existing = await prisma_1.prisma.room.findUnique({ where: { name } });
    if (existing)
        throw (0, error_middleware_1.createError)('Room name already taken', 409);
    return prisma_1.prisma.room.create({
        data: {
            name, description, visibility, ownerId,
            members: { create: { userId: ownerId, role: client_1.MemberRole.ADMIN } },
        },
        select: roomSelect,
    });
}
async function getPublicRooms(search) {
    return prisma_1.prisma.room.findMany({
        where: {
            visibility: client_1.RoomVisibility.PUBLIC,
            isDirect: false,
            ...(search ? { name: { contains: search, mode: 'insensitive' } } : {}),
        },
        select: roomSelect,
        orderBy: { createdAt: 'desc' },
    });
}
async function getUserRooms(userId) {
    const memberships = await prisma_1.prisma.roomMember.findMany({
        where: { userId },
        include: {
            room: { select: roomSelect },
        },
    });
    return memberships.map(m => ({ ...m.room, role: m.role }));
}
async function getRoomById(roomId, userId) {
    const member = await prisma_1.prisma.roomMember.findUnique({
        where: { roomId_userId: { roomId, userId } },
    });
    if (!member)
        throw (0, error_middleware_1.createError)('Not a member of this room', 403);
    return prisma_1.prisma.room.findUnique({
        where: { id: roomId },
        select: {
            ...roomSelect,
            members: {
                include: { user: { select: { id: true, username: true } } },
            },
        },
    });
}
async function joinRoom(roomId, userId) {
    const room = await prisma_1.prisma.room.findUnique({ where: { id: roomId } });
    if (!room)
        throw (0, error_middleware_1.createError)('Room not found', 404);
    if (room.visibility === client_1.RoomVisibility.PRIVATE)
        throw (0, error_middleware_1.createError)('Room is private', 403);
    const ban = await prisma_1.prisma.roomBan.findUnique({
        where: { roomId_userId: { roomId, userId } },
    });
    if (ban)
        throw (0, error_middleware_1.createError)('You are banned from this room', 403);
    const existing = await prisma_1.prisma.roomMember.findUnique({
        where: { roomId_userId: { roomId, userId } },
    });
    if (existing)
        throw (0, error_middleware_1.createError)('Already a member', 400);
    await prisma_1.prisma.roomMember.create({ data: { roomId, userId } });
    return room;
}
async function leaveRoom(roomId, userId) {
    const room = await prisma_1.prisma.room.findUnique({ where: { id: roomId } });
    if (!room)
        throw (0, error_middleware_1.createError)('Room not found', 404);
    if (room.ownerId === userId)
        throw (0, error_middleware_1.createError)('Owner cannot leave. Delete the room instead.', 400);
    await prisma_1.prisma.roomMember.delete({
        where: { roomId_userId: { roomId, userId } },
    });
}
async function deleteRoom(roomId, userId) {
    const room = await prisma_1.prisma.room.findUnique({ where: { id: roomId } });
    if (!room)
        throw (0, error_middleware_1.createError)('Room not found', 404);
    if (room.ownerId !== userId)
        throw (0, error_middleware_1.createError)('Only the owner can delete the room', 403);
    await prisma_1.prisma.room.delete({ where: { id: roomId } });
}
async function updateRoom(roomId, userId, data) {
    const room = await prisma_1.prisma.room.findUnique({ where: { id: roomId } });
    if (!room)
        throw (0, error_middleware_1.createError)('Room not found', 404);
    if (room.ownerId !== userId)
        throw (0, error_middleware_1.createError)('Only the owner can update the room', 403);
    if (data.name && data.name !== room.name) {
        const existing = await prisma_1.prisma.room.findUnique({ where: { name: data.name } });
        if (existing)
            throw (0, error_middleware_1.createError)('Room name already taken', 409);
    }
    return prisma_1.prisma.room.update({ where: { id: roomId }, data, select: roomSelect });
}
async function getRoomMembers(roomId, userId) {
    const member = await prisma_1.prisma.roomMember.findUnique({
        where: { roomId_userId: { roomId, userId } },
    });
    if (!member)
        throw (0, error_middleware_1.createError)('Not a member', 403);
    return prisma_1.prisma.roomMember.findMany({
        where: { roomId },
        include: { user: { select: { id: true, username: true } } },
    });
}
async function banMember(roomId, targetUserId, adminId) {
    const room = await prisma_1.prisma.room.findUnique({ where: { id: roomId } });
    if (!room)
        throw (0, error_middleware_1.createError)('Room not found', 404);
    const admin = await prisma_1.prisma.roomMember.findUnique({
        where: { roomId_userId: { roomId, userId: adminId } },
    });
    if (!admin || admin.role !== client_1.MemberRole.ADMIN)
        throw (0, error_middleware_1.createError)('Not authorized', 403);
    if (room.ownerId === targetUserId)
        throw (0, error_middleware_1.createError)('Cannot ban the owner', 400);
    await prisma_1.prisma.$transaction([
        prisma_1.prisma.roomMember.deleteMany({ where: { roomId, userId: targetUserId } }),
        prisma_1.prisma.roomBan.upsert({
            where: { roomId_userId: { roomId, userId: targetUserId } },
            create: { roomId, userId: targetUserId, bannedById: adminId },
            update: { bannedById: adminId },
        }),
    ]);
}
async function unbanMember(roomId, targetUserId, adminId) {
    const admin = await prisma_1.prisma.roomMember.findUnique({
        where: { roomId_userId: { roomId, userId: adminId } },
    });
    if (!admin || admin.role !== client_1.MemberRole.ADMIN)
        throw (0, error_middleware_1.createError)('Not authorized', 403);
    await prisma_1.prisma.roomBan.delete({
        where: { roomId_userId: { roomId, userId: targetUserId } },
    });
}
async function getBannedUsers(roomId, adminId) {
    const admin = await prisma_1.prisma.roomMember.findUnique({
        where: { roomId_userId: { roomId, userId: adminId } },
    });
    if (!admin || admin.role !== client_1.MemberRole.ADMIN)
        throw (0, error_middleware_1.createError)('Not authorized', 403);
    return prisma_1.prisma.roomBan.findMany({
        where: { roomId },
        include: {
            user: { select: { id: true, username: true } },
            bannedBy: { select: { id: true, username: true } },
        },
    });
}
async function makeAdmin(roomId, targetUserId, ownerId) {
    const room = await prisma_1.prisma.room.findUnique({ where: { id: roomId } });
    if (!room || room.ownerId !== ownerId)
        throw (0, error_middleware_1.createError)('Only owner can manage admins', 403);
    await prisma_1.prisma.roomMember.update({
        where: { roomId_userId: { roomId, userId: targetUserId } },
        data: { role: client_1.MemberRole.ADMIN },
    });
}
async function removeAdmin(roomId, targetUserId, ownerId) {
    const room = await prisma_1.prisma.room.findUnique({ where: { id: roomId } });
    if (!room || room.ownerId !== ownerId)
        throw (0, error_middleware_1.createError)('Only owner can manage admins', 403);
    if (targetUserId === ownerId)
        throw (0, error_middleware_1.createError)('Cannot remove owner admin rights', 400);
    await prisma_1.prisma.roomMember.update({
        where: { roomId_userId: { roomId, userId: targetUserId } },
        data: { role: client_1.MemberRole.MEMBER },
    });
}
async function removeMember(roomId, targetUserId, adminId) {
    await banMember(roomId, targetUserId, adminId);
}
async function inviteUser(roomId, inviterId, inviteeUsername) {
    const room = await prisma_1.prisma.room.findUnique({ where: { id: roomId } });
    if (!room)
        throw (0, error_middleware_1.createError)('Room not found', 404);
    const member = await prisma_1.prisma.roomMember.findUnique({
        where: { roomId_userId: { roomId, userId: inviterId } },
    });
    if (!member)
        throw (0, error_middleware_1.createError)('Not a member', 403);
    const invitee = await prisma_1.prisma.user.findUnique({ where: { username: inviteeUsername } });
    if (!invitee)
        throw (0, error_middleware_1.createError)('User not found', 404);
    // Check if already banned
    const ban = await prisma_1.prisma.roomBan.findUnique({
        where: { roomId_userId: { roomId, userId: invitee.id } },
    });
    if (ban)
        throw (0, error_middleware_1.createError)('User is banned from this room', 400);
    // If public room, just add them
    if (room.visibility === client_1.RoomVisibility.PUBLIC) {
        const existing = await prisma_1.prisma.roomMember.findUnique({
            where: { roomId_userId: { roomId, userId: invitee.id } },
        });
        if (!existing) {
            await prisma_1.prisma.roomMember.create({ data: { roomId, userId: invitee.id } });
        }
        return;
    }
    await prisma_1.prisma.roomInvitation.upsert({
        where: { roomId_inviteeId: { roomId, inviteeId: invitee.id } },
        create: { roomId, inviterId, inviteeId: invitee.id },
        update: {},
    });
    return invitee;
}
async function acceptInvitation(invitationId, userId) {
    const invitation = await prisma_1.prisma.roomInvitation.findUnique({ where: { id: invitationId } });
    if (!invitation || invitation.inviteeId !== userId)
        throw (0, error_middleware_1.createError)('Invitation not found', 404);
    await prisma_1.prisma.$transaction([
        prisma_1.prisma.roomMember.upsert({
            where: { roomId_userId: { roomId: invitation.roomId, userId } },
            create: { roomId: invitation.roomId, userId },
            update: {},
        }),
        prisma_1.prisma.roomInvitation.delete({ where: { id: invitationId } }),
    ]);
}
async function getPendingInvitations(userId) {
    return prisma_1.prisma.roomInvitation.findMany({
        where: { inviteeId: userId },
        include: {
            room: { select: { id: true, name: true, description: true } },
            inviter: { select: { id: true, username: true } },
        },
    });
}
// Creates or gets a DM room between two users
async function getOrCreateDM(userId1, userId2) {
    // DM rooms have a deterministic name
    const [a, b] = [userId1, userId2].sort();
    const dmName = `dm:${a}:${b}`;
    const existing = await prisma_1.prisma.room.findUnique({ where: { name: dmName } });
    if (existing)
        return existing;
    return prisma_1.prisma.room.create({
        data: {
            name: dmName,
            visibility: client_1.RoomVisibility.PRIVATE,
            ownerId: userId1,
            isDirect: true,
            members: {
                create: [
                    { userId: userId1, role: client_1.MemberRole.MEMBER },
                    { userId: userId2, role: client_1.MemberRole.MEMBER },
                ],
            },
        },
        select: roomSelect,
    });
}
