// chat-backend/dist/modules/presence/presence.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerTab = registerTab;
exports.unregisterTab = unregisterTab;
exports.heartbeat = heartbeat;
exports.recalculatePresence = recalculatePresence;
exports.setPresence = setPresence;
exports.getPresence = getPresence;
exports.getBulkPresence = getBulkPresence;
const redis_1 = require("../../lib/redis");
const AFK_THRESHOLD_MS = 60 * 1000; // 1 minute
async function registerTab(userId, tabId) {
    await redis_1.redis.sadd((0, redis_1.tabKey)(userId), tabId);
    await redis_1.redis.set((0, redis_1.lastActiveKey)(userId, tabId), Date.now().toString());
    await setPresence(userId, 'online');
}
async function unregisterTab(userId, tabId) {
    await redis_1.redis.srem((0, redis_1.tabKey)(userId), tabId);
    await redis_1.redis.del((0, redis_1.lastActiveKey)(userId, tabId));
    const remaining = await redis_1.redis.smembers((0, redis_1.tabKey)(userId));
    if (remaining.length === 0) {
        await setPresence(userId, 'offline');
    }
    else {
        await recalculatePresence(userId);
    }
}
async function heartbeat(userId, tabId) {
    await redis_1.redis.set((0, redis_1.lastActiveKey)(userId, tabId), Date.now().toString());
    await recalculatePresence(userId);
}
async function recalculatePresence(userId) {
    const tabs = await redis_1.redis.smembers((0, redis_1.tabKey)(userId));
    if (tabs.length === 0) {
        await setPresence(userId, 'offline');
        return 'offline';
    }
    const now = Date.now();
    let anyActive = false;
    for (const tabId of tabs) {
        const lastActive = await redis_1.redis.get((0, redis_1.lastActiveKey)(userId, tabId));
        if (lastActive && now - parseInt(lastActive, 10) < AFK_THRESHOLD_MS) {
            anyActive = true;
            break;
        }
    }
    const status = anyActive ? 'online' : 'afk';
    await setPresence(userId, status);
    return status;
}
async function setPresence(userId, status) {
    await redis_1.redis.set((0, redis_1.presenceKey)(userId), status, 'EX', 3600);
}
async function getPresence(userId) {
    const status = await redis_1.redis.get((0, redis_1.presenceKey)(userId));
    return status || 'offline';
}
async function getBulkPresence(userIds) {
    if (userIds.length === 0)
        return {};
    const keys = userIds.map(redis_1.presenceKey);
    const values = await redis_1.redis.mget(...keys);
    return userIds.reduce((acc, userId, i) => {
        acc[userId] = values[i] || 'offline';
        return acc;
    }, {});
}
