// chat-backend/dist/modules/uploads/uploads.routes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const express_1 = require("express");
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const auth_middleware_1 = require("../../common/middleware/auth.middleware");
const upload_middleware_1 = require("../../common/middleware/upload.middleware");
const prisma_1 = require("../../lib/prisma");
const env_1 = require("../../config/env");
const error_middleware_1 = require("../../common/middleware/error.middleware");
const router = (0, express_1.Router)();
router.use(auth_middleware_1.authMiddleware);
const wrap = (fn) => (req, res, next) => fn(req, res, next);
// Upload image attached to a message
router.post('/image/:roomId', upload_middleware_1.uploadImage.single('file'), wrap(async (req, res, next) => {
    try {
        if (!req.file)
            throw (0, error_middleware_1.createError)('No file uploaded', 400);
        const member = await prisma_1.prisma.roomMember.findUnique({
            where: { roomId_userId: { roomId: req.params.roomId, userId: req.userId } },
        });
        if (!member)
            throw (0, error_middleware_1.createError)('Not a member', 403);
        const { comment, messageId } = req.body;
        let targetMessageId = messageId;
        if (!targetMessageId) {
            const msg = await prisma_1.prisma.message.create({
                data: { roomId: req.params.roomId, senderId: req.userId, content: comment || null },
            });
            targetMessageId = msg.id;
        }
        const attachment = await prisma_1.prisma.attachment.create({
            data: {
                messageId: targetMessageId,
                filename: req.file.filename,
                originalName: req.file.originalname,
                size: req.file.size,
                mimeType: req.file.mimetype,
                comment,
            },
        });
        res.status(201).json({ attachment, messageId: targetMessageId });
    }
    catch (err) {
        next(err);
    }
}));
// Upload arbitrary file
router.post('/file/:roomId', upload_middleware_1.uploadFile.single('file'), wrap(async (req, res, next) => {
    try {
        if (!req.file)
            throw (0, error_middleware_1.createError)('No file uploaded', 400);
        const member = await prisma_1.prisma.roomMember.findUnique({
            where: { roomId_userId: { roomId: req.params.roomId, userId: req.userId } },
        });
        if (!member)
            throw (0, error_middleware_1.createError)('Not a member', 403);
        const { comment, messageId } = req.body;
        let targetMessageId = messageId;
        if (!targetMessageId) {
            const msg = await prisma_1.prisma.message.create({
                data: { roomId: req.params.roomId, senderId: req.userId, content: comment || null },
            });
            targetMessageId = msg.id;
        }
        const attachment = await prisma_1.prisma.attachment.create({
            data: {
                messageId: targetMessageId,
                filename: req.file.filename,
                originalName: req.file.originalname,
                size: req.file.size,
                mimeType: req.file.mimetype,
                comment,
            },
        });
        res.status(201).json({ attachment, messageId: targetMessageId });
    }
    catch (err) {
        next(err);
    }
}));
// Download file — access controlled
router.get('/:type/:filename', wrap(async (req, res, next) => {
    try {
        const { type, filename } = req.params;
        if (!['images', 'files'].includes(type))
            throw (0, error_middleware_1.createError)('Invalid file type', 400);
        // Find attachment
        const attachment = await prisma_1.prisma.attachment.findFirst({ where: { filename } });
        if (!attachment)
            throw (0, error_middleware_1.createError)('File not found', 404);
        // Check membership
        const message = await prisma_1.prisma.message.findUnique({ where: { id: attachment.messageId } });
        if (!message)
            throw (0, error_middleware_1.createError)('File not found', 404);
        const member = await prisma_1.prisma.roomMember.findUnique({
            where: { roomId_userId: { roomId: message.roomId, userId: req.userId } },
        });
        if (!member)
            throw (0, error_middleware_1.createError)('Access denied', 403);
        const filePath = path_1.default.join(process.cwd(), env_1.config.uploadsDir, type, filename);
        if (!fs_1.default.existsSync(filePath))
            throw (0, error_middleware_1.createError)('File not found on disk', 404);
        res.setHeader('Content-Disposition', `attachment; filename="${attachment.originalName}"`);
        res.sendFile(path_1.default.resolve(filePath));
    }
    catch (err) {
        next(err);
    }
}));
exports.default = router;
