// chat-backend/dist/socket/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initSocket = initSocket;
const socket_io_1 = require("socket.io");
const env_1 = require("../config/env");
const jwt_1 = require("../common/utils/jwt");
const prisma_1 = require("../lib/prisma");
const presence_service_1 = require("../modules/presence/presence.service");
const messages_service_1 = require("../modules/messages/messages.service");
const uuid_1 = require("uuid");
function initSocket(httpServer) {
    const io = new socket_io_1.Server(httpServer, {
        cors: {
            origin: env_1.config.clientUrl,
            credentials: true,
        },
        pingInterval: 25000,
        pingTimeout: 60000,
    });
    // Authentication middleware
    io.use(async (socket, next) => {
        try {
            const token = socket.handshake.auth.token || socket.handshake.headers.cookie
                ?.split(';').find((c) => c.trim().startsWith('token='))
                ?.split('=')[1];
            if (!token)
                return next(new Error('Unauthorized'));
            const payload = (0, jwt_1.verifyToken)(token);
            const session = await prisma_1.prisma.session.findUnique({
                where: { id: payload.sessionId, token },
            });
            if (!session || session.expiresAt < new Date()) {
                return next(new Error('Session expired'));
            }
            socket.userId = payload.userId;
            socket.sessionId = payload.sessionId;
            socket.tabId = (0, uuid_1.v4)();
            next();
        }
        catch {
            next(new Error('Unauthorized'));
        }
    });
    io.on('connection', async (rawSocket) => {
        const socket = rawSocket;
        const { userId, tabId } = socket;
        console.log(`User ${userId} connected (tab ${tabId})`);
        // Register tab and broadcast presence
        await (0, presence_service_1.registerTab)(userId, tabId);
        socket.join(`user:${userId}`);
        // Join all user's rooms
        const memberships = await prisma_1.prisma.roomMember.findMany({
            where: { userId },
            select: { roomId: true },
        });
        for (const { roomId } of memberships) {
            socket.join(`room:${roomId}`);
        }
        // Broadcast online status to contacts
        await broadcastPresence(io, userId, 'online');
        // Heartbeat — client sends every 30s
        socket.on('heartbeat', async () => {
            await (0, presence_service_1.heartbeat)(userId, tabId);
            const presence = await (0, presence_service_1.getBulkPresence)([userId]);
            if (presence[userId] !== 'online') {
                await broadcastPresence(io, userId, presence[userId]);
            }
        });
        // Activity ping — resets AFK timer
        socket.on('activity', async () => {
            await (0, presence_service_1.heartbeat)(userId, tabId);
        });
        // --- MESSAGING ---
        socket.on('message:send', async (data, ack) => {
            try {
                const message = await (0, messages_service_1.sendMessage)(data.roomId, userId, data.content, data.replyToId);
                io.to(`room:${data.roomId}`).emit('message:new', message);
                ack?.({ ok: true, message });
            }
            catch (err) {
                ack?.({ ok: false, error: err.message });
            }
        });
        socket.on('message:edit', async (data, ack) => {
            try {
                const message = await (0, messages_service_1.editMessage)(data.messageId, userId, data.content);
                io.to(`room:${message.roomId}`).emit('message:updated', message);
                ack?.({ ok: true, message });
            }
            catch (err) {
                ack?.({ ok: false, error: err.message });
            }
        });
        socket.on('message:delete', async (data, ack) => {
            try {
                await (0, messages_service_1.deleteMessage)(data.messageId, userId);
                io.to(`room:${data.roomId}`).emit('message:deleted', { messageId: data.messageId, roomId: data.roomId });
                ack?.({ ok: true });
            }
            catch (err) {
                ack?.({ ok: false, error: err.message });
            }
        });
        // --- ROOM EVENTS ---
        socket.on('room:join', (roomId) => {
            socket.join(`room:${roomId}`);
        });
        socket.on('room:leave', (roomId) => {
            socket.leave(`room:${roomId}`);
        });
        // Typing indicators
        socket.on('typing:start', (roomId) => {
            socket.to(`room:${roomId}`).emit('typing:start', { userId, roomId });
        });
        socket.on('typing:stop', (roomId) => {
            socket.to(`room:${roomId}`).emit('typing:stop', { userId, roomId });
        });
        // --- DISCONNECT ---
        socket.on('disconnect', async () => {
            console.log(`User ${userId} disconnected (tab ${tabId})`);
            await (0, presence_service_1.unregisterTab)(userId, tabId);
            const presence = await (0, presence_service_1.getBulkPresence)([userId]);
            await broadcastPresence(io, userId, presence[userId]);
        });
    });
    // AFK check interval — every 30 seconds
    setInterval(async () => {
        const sockets = await io.fetchSockets();
        const userIds = [...new Set(sockets.map(s => s.userId))];
        if (userIds.length === 0)
            return;
        const presenceMap = await (0, presence_service_1.getBulkPresence)(userIds);
        for (const [uid, status] of Object.entries(presenceMap)) {
            io.to(`user:${uid}`).emit('presence:self', { status });
        }
    }, 30000);
    return io;
}
async function broadcastPresence(io, userId, status) {
    // Get user's contacts and room members to notify
    const friendships = await prisma_1.prisma.friendship.findMany({
        where: {
            status: 'ACCEPTED',
            OR: [{ requesterId: userId }, { addresseeId: userId }],
        },
        select: { requesterId: true, addresseeId: true },
    });
    const contactIds = friendships.map(f => f.requesterId === userId ? f.addresseeId : f.requesterId);
    // Notify contacts
    for (const contactId of contactIds) {
        io.to(`user:${contactId}`).emit('presence:update', { userId, status });
    }
    // Notify room members
    const memberships = await prisma_1.prisma.roomMember.findMany({
        where: { userId },
        select: { roomId: true },
    });
    for (const { roomId } of memberships) {
        io.to(`room:${roomId}`).emit('presence:update', { userId, status });
    }
}
