// chat-backend/dist/modules/friends/friends.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendFriendRequest = sendFriendRequest;
exports.respondToRequest = respondToRequest;
exports.getFriends = getFriends;
exports.getPendingRequests = getPendingRequests;
exports.removeFriend = removeFriend;
exports.banUser = banUser;
exports.unbanUser = unbanUser;
exports.getBannedUsers = getBannedUsers;
exports.canMessage = canMessage;
const prisma_1 = require("../../lib/prisma");
const error_middleware_1 = require("../../common/middleware/error.middleware");
const client_1 = require("@prisma/client");
async function sendFriendRequest(requesterId, addresseeUsername, message) {
    const addressee = await prisma_1.prisma.user.findUnique({ where: { username: addresseeUsername } });
    if (!addressee)
        throw (0, error_middleware_1.createError)('User not found', 404);
    if (addressee.id === requesterId)
        throw (0, error_middleware_1.createError)('Cannot add yourself', 400);
    // Check ban
    const ban = await prisma_1.prisma.userBan.findUnique({
        where: { bannerId_bannedId: { bannerId: addressee.id, bannedId: requesterId } },
    });
    if (ban)
        throw (0, error_middleware_1.createError)('Cannot send request to this user', 403);
    const existing = await prisma_1.prisma.friendship.findFirst({
        where: {
            OR: [
                { requesterId, addresseeId: addressee.id },
                { requesterId: addressee.id, addresseeId: requesterId },
            ],
        },
    });
    if (existing)
        throw (0, error_middleware_1.createError)('Friend request already exists or already friends', 400);
    return prisma_1.prisma.friendship.create({
        data: { requesterId, addresseeId: addressee.id, message },
        include: {
            addressee: { select: { id: true, username: true } },
        },
    });
}
async function respondToRequest(friendshipId, userId, accept) {
    const friendship = await prisma_1.prisma.friendship.findUnique({ where: { id: friendshipId } });
    if (!friendship || friendship.addresseeId !== userId)
        throw (0, error_middleware_1.createError)('Request not found', 404);
    if (friendship.status !== client_1.FriendshipStatus.PENDING)
        throw (0, error_middleware_1.createError)('Request already handled', 400);
    if (!accept) {
        await prisma_1.prisma.friendship.delete({ where: { id: friendshipId } });
        return null;
    }
    return prisma_1.prisma.friendship.update({
        where: { id: friendshipId },
        data: { status: client_1.FriendshipStatus.ACCEPTED },
        include: {
            requester: { select: { id: true, username: true } },
            addressee: { select: { id: true, username: true } },
        },
    });
}
async function getFriends(userId) {
    const friendships = await prisma_1.prisma.friendship.findMany({
        where: {
            status: client_1.FriendshipStatus.ACCEPTED,
            OR: [{ requesterId: userId }, { addresseeId: userId }],
        },
        include: {
            requester: { select: { id: true, username: true } },
            addressee: { select: { id: true, username: true } },
        },
    });
    return friendships.map(f => {
        const friend = f.requesterId === userId ? f.addressee : f.requester;
        return { friendshipId: f.id, friend };
    });
}
async function getPendingRequests(userId) {
    return prisma_1.prisma.friendship.findMany({
        where: { addresseeId: userId, status: client_1.FriendshipStatus.PENDING },
        include: {
            requester: { select: { id: true, username: true } },
        },
    });
}
async function removeFriend(userId, friendId) {
    const friendship = await prisma_1.prisma.friendship.findFirst({
        where: {
            status: client_1.FriendshipStatus.ACCEPTED,
            OR: [
                { requesterId: userId, addresseeId: friendId },
                { requesterId: friendId, addresseeId: userId },
            ],
        },
    });
    if (!friendship)
        throw (0, error_middleware_1.createError)('Not friends', 404);
    await prisma_1.prisma.friendship.delete({ where: { id: friendship.id } });
}
async function banUser(bannerId, bannedUsername) {
    const banned = await prisma_1.prisma.user.findUnique({ where: { username: bannedUsername } });
    if (!banned)
        throw (0, error_middleware_1.createError)('User not found', 404);
    if (banned.id === bannerId)
        throw (0, error_middleware_1.createError)('Cannot ban yourself', 400);
    // Remove friendship if exists
    await prisma_1.prisma.friendship.deleteMany({
        where: {
            OR: [
                { requesterId: bannerId, addresseeId: banned.id },
                { requesterId: banned.id, addresseeId: bannerId },
            ],
        },
    });
    await prisma_1.prisma.userBan.upsert({
        where: { bannerId_bannedId: { bannerId, bannedId: banned.id } },
        create: { bannerId, bannedId: banned.id },
        update: {},
    });
}
async function unbanUser(bannerId, bannedId) {
    await prisma_1.prisma.userBan.delete({
        where: { bannerId_bannedId: { bannerId, bannedId } },
    });
}
async function getBannedUsers(userId) {
    return prisma_1.prisma.userBan.findMany({
        where: { bannerId: userId },
        include: { banned: { select: { id: true, username: true } } },
    });
}
async function canMessage(userId1, userId2) {
    const friendship = await prisma_1.prisma.friendship.findFirst({
        where: {
            status: client_1.FriendshipStatus.ACCEPTED,
            OR: [
                { requesterId: userId1, addresseeId: userId2 },
                { requesterId: userId2, addresseeId: userId1 },
            ],
        },
    });
    if (!friendship)
        return false;
    const ban = await prisma_1.prisma.userBan.findFirst({
        where: {
            OR: [
                { bannerId: userId1, bannedId: userId2 },
                { bannerId: userId2, bannedId: userId1 },
            ],
        },
    });
    return !ban;
}
